pkg/api/steve/supportconfigs/generator.go

```diff
@@ -0,0 +1,166 @@
+		return
+	}
+	logrus.Infof("[%s] Generating supportconfig", logPrefix)
+	archive, err := h.generateSupportConfig()
+	logrus.Infof("[%s] Done Generating supportconfig", logPrefix)
+	if err != nil {
+		if errors.Is(err, errNotImplemented) {
+			util.ReturnHTTPError(writer, request, http.StatusNotImplemented, "csp-adapter must be installed in order to to generate supportconfigs")
+			return
+		}
+		logrus.Errorf("[%s] Error when generating supportconfig %v", logPrefix, err)
+		util.ReturnHTTPError(writer, request, http.StatusInternalServerError, "Internal Server Error")
+		return
+	}
+	writer.Header().Set("Content-Type", tarContentType)
+	writer.Header().Set("Content-Disposition", "attachment; filename=\"supportconfig.tar\"")
+	n, err := io.Copy(writer, archive)
+	logrus.Debugf("[%s] wrote %v bytes in archive response", logPrefix, n)
+}
+
+// authorize checks to see if the user can get the csp adapter configmap. Returns a bool (if the user is authorized)
+// and optionally an error
+func (h *GeneratorHandler) authorize(r *http.Request) (bool, error) {
+	userInfo, ok := request.UserFrom(r.Context())
+	if !ok {
+		return false, fmt.Errorf("unable to extract user info from context")
+	}
+	response, err := h.SubjectAccessReviews.Create(r.Context(), &authzv1.SubjectAccessReview{
+		Spec: authzv1.SubjectAccessReviewSpec{
+			ResourceAttributes: &authzv1.ResourceAttributes{
+				Resource:  "configmap",
+				Verb:      "get",
+				Name:      cspAdapterConfigmap,
+				Namespace: cspAdapterNamespace,
+			},
+			User:   userInfo.GetName(),
+			Groups: userInfo.GetGroups(),
+			Extra:  convertExtra(userInfo.GetExtra()),
+			UID:    userInfo.GetUID(),
+		},
+	}, metav1.CreateOptions{})
+	if err != nil {
+		return false, fmt.Errorf("failed to create sar %s", err)
+	}
+	if !response.Status.Allowed {
+		return false, nil
+	}
+	return true, nil
+}
+
+// generateSupportConfig produces an io.Reader with the supportconfig ready to be returned using a http.ResponseWriter
+// Returns an err if it can't get the support config (note: does not return an err if it can't find the csp config)
+func (h *GeneratorHandler) generateSupportConfig() (io.Reader, error) {
+	cspConfig, err := h.getCspConfig()
+	// csp adapter won't always be installed
+	if err != nil {
+		return nil, err
+	}
+	configData, err := json.MarshalIndent(cspConfig, "", "  ")
+	if err != nil {
+		return nil, err
+	}
+
+	var buf bytes.Buffer
+	tw := tar.NewWriter(&buf)
+
+	files := map[string][]byte{
+		"rancher/config.json": configData,
+	}
+
+	for name, body := range files {
+		hdr := &tar.Header{
+			Name: name,
+			Mode: 0600,
+			Size: int64(len(body)),
+		}
+		if err := tw.WriteHeader(hdr); err != nil {
+			return nil, err
+		}
+		if _, err := tw.Write(body); err != nil {
+			return nil, err
+		}
+	}
+
+	return &buf, nil
+}
+
+// getCspConfig gets the configmap produced by the csp-adapter
+func (h *GeneratorHandler) getCspConfig() (map[string]interface{}, error) {
```

Comment:
nit: with acronyms as CSP I think it makes sense to keep them all the same case, here that would be upper.